flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DefaultResourceAllocationStrategy.java

```diff
@@ -142,115 +145,98 @@ private static ResourceCounter tryFulfillRequirementsForJobWithRegisteredResourc
                     getEffectiveResourceProfile(
                             missingResource.getKey(), defaultSlotResourceProfile);
-            for (int i = 0; i < missingResource.getValue(); i++) {
-                Optional<PendingTaskManagerId> matchedPendingTaskManager =
-                        findPendingManagerToFulfill(effectiveProfile, availableResources);
-                if (matchedPendingTaskManager.isPresent()) {
-                    availableResources.compute(
-                            matchedPendingTaskManager.get(),
-                            ((pendingTaskManagerId, resourceProfile) ->
-                                    Preconditions.checkNotNull(resourceProfile)
-                                            .subtract(effectiveProfile)));
+            final Iterator<Tuple2<PendingTaskManagerId, ResourceProfile>> availableResourcesItr =
+                    availableResources.iterator();
+            int numUnfulfilled = missingResource.getValue();
+            while (numUnfulfilled > 0 && availableResourcesItr.hasNext()) {
+                final Tuple2<PendingTaskManagerId, ResourceProfile> pendingTaskManager =
+                        availableResourcesItr.next();
+                while (numUnfulfilled > 0
+                        && canFulfillRequirement(effectiveProfile, pendingTaskManager.f1)) {
                     resultBuilder.addAllocationOnPendingResource(
-                            jobId, matchedPendingTaskManager.get(), effectiveProfile);
-                } else {
-                    if (totalResourceProfile.allFieldsNoLessThan(effectiveProfile)) {
-                        // Add new pending task manager
-                        final PendingTaskManager pendingTaskManager =
-                                new PendingTaskManager(totalResourceProfile, numSlotsPerWorker);
-                        resultBuilder.addPendingTaskManagerAllocate(pendingTaskManager);
-                        resultBuilder.addAllocationOnPendingResource(
-                                jobId,
-                                pendingTaskManager.getPendingTaskManagerId(),
-                                effectiveProfile);
-                        availableResources.put(
-                                pendingTaskManager.getPendingTaskManagerId(),
-                                totalResourceProfile.subtract(effectiveProfile));
-                    } else {
-                        resultBuilder.addUnfulfillableJob(jobId);
-                        break;
-                    }
+                            jobId, pendingTaskManager.f0, effectiveProfile);
+                    numUnfulfilled--;
+                    pendingTaskManager.setField(
+                            pendingTaskManager.f1.subtract(effectiveProfile), 1);
+                }
+                if (pendingTaskManager.f1.equals(ResourceProfile.ZERO)) {
+                    availableResourcesItr.remove();
+                }
+            }
```

Comment:
I wonder if it make sense to further deduplicate the matching logics for registered/pending resources. As far as I can see, they are very similar. The only differences are:
- Action to take when resource and requirement are matched. We can make it a function parameter.
- Handling of unfulfilled requirements (allocating more pending TMs and report unfulfillable job), which is the very last step.